8326936: RISC-V: Shenandoah GC crashes due to incorrect atomic memory operations

MaxXSoft opened the pull request:
With compressed oops enabled, Shenandoah GC crashes in the concurrent marking phase due to some incorrect atomic memory operations. This resulted in the failure of multiple related tests, including `gc/shenandoah/TestSmallHeap.java`, `gc/metaspace/TestMetaspacePerfCounters.java#Shenandoah-64`, and so on, tested on XuanTie C910 and QEMU.

This failure is related to a GCC bug we recently discovered: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=114130.

In detail, there's a pattern in method `ShenandoahHeap::conc_update_with_forwarded`:

```cpp
if (in_collection_set(obj)) {
  // Dereferences `obj` without explicit null check.
  oop fwd = ShenandoahBarrierSet::resolve_forwarded_not_null(obj);
  // Then calls atomic built-in.
  atomic_update_oop(fwd, p, obj);
}
```

`atomic_update_oop` then compresses `obj` into a 32-bit word and calls the built-in `__atomic_compare_exchange`. The latter produces incorrect assembly that comparing this unsigned 32-bit word with the sign-extended result of `lr.w` instructions.

This bug can be bypassed by adding an explicit null check (like `if (obj && in_collection_set(obj))`), or adding compiler flag `-fno-delete-null-pointer-checks`.

In previous commits, JDK-8316186 removed RISC-V's `PlatformCmpxchg<4>` but left the flag `-fno-delete-null-pointer-checks` enabled. Then JDK-8316893 removed the flag and made the bug visible. This patch adds `PlatformCmpxchg<4>` back.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326936](https://bugs.openjdk.org/browse/JDK-8326936): RISC-V: Shenandoah GC crashes due to incorrect atomic memory operations (**Bug** - P4)


### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [e1a49d85](https://git.openjdk.org/jdk/pull/18039/files/e1a49d8530e4e059ecea28c376d6b9d82eee2b4e)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18039/head:pull/18039` \
`$ git checkout pull/18039`

Update a local copy of the PR: \
`$ git checkout pull/18039` \
`$ git pull https://git.openjdk.org/jdk.git pull/18039/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18039`

View PR using the GUI difftool: \
`$ git pr show -t 18039`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18039.diff">https://git.openjdk.org/jdk/pull/18039.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18039#issuecomment-1968630980)

RealFYang commented:
Hi, I don't witness such an issue when running those regression tests with jdk binaries built with native GCC-12 on Licheepi-4a or Unmatched boards. Do you know why? Does this only trigger for jdk binaries built with GCC-13? Guess @robehn who owns https://bugs.openjdk.org/browse/JDK-8316186 might also be interested in this.

robehn commented:
Hi, on vacation, back next week.

In this case old_value is unsigned, but in the assembly we loaded a sign extended value.
Return value thus seems always singed extended, as we plainly return old_value.
As the compiler can't know it should zero extend an 'uint32_t' on return.
So this does not look correct to me?

Anyhow can't we just use a int32_t when comparing, and cast that back to the 'uint32_t' on return? (and still use __atomic_...)

Sorry not really available, as I said back next week.
Thanks for finding!

MaxXSoft commented:
Hello, @robehn !

Thanks for your reply and suggestion. I tried this before, and this work around does not work. That's why I reported this bug to GCC, it's so weird.

> In this case old_value is unsigned, but in the assembly we loaded a sign extended value.
> Return value thus seems always singed extended, as we plainly return old_value.
> As the compiler can't know it should zero extend an 'uint32_t' on return.
> So this does not look correct to me?

According to what I observed, it's correct. [RISC-V calling convention](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc) says:

> When passed in registers or on the stack, integer scalars narrower than XLEN bits are widened according to the sign of their type up to 32 bits, then sign-extended to XLEN bits.

And then it says:

> Values are returned in the same manner as a first named argument of the same type would be passed.

So the return value should be sign-extended, even if it's unsigned (according to my understand). Here's a simple example: https://godbolt.org/z/n6vfoYPPx. GCC generates `slliw` for the unsigned return value.

> Anyhow can't we just use a int32_t when comparing, and cast that back to the 'uint32_t' on return? (and still use _atomic...)

Seems we can't. Check this example for more details: https://godbolt.org/z/cv5ejTfxh. I wrote both the inline assembly version and the built-in version CAS, and tested them on the same code snippet. The inline assembly version works well due to the ABI, but the built-in version is still buggy.

I also tried to build JDK with built-in atomic with a unsigned to signed conversion as the above example, and unsurprisingly it failed on Shenandoah-related tests.

In addition I removed the redundant `sext.w` in `PlatformCmpxchg<4>`, it works well. Dump of `libjvm.so`:

```
  ab2d04:	02a88663          	beq	a7,a0,ab2d30 <_ZN14ShenandoahHeap26conc_update_with_forwardedI9narrowOopEEvPT_+0x7c>
  ab2d08:	00c7d7b3          	srl	a5,a5,a2
  ab2d0c:	00c75733          	srl	a4,a4,a2
  ab2d10:	2781                	sext.w	a5,a5
  ab2d12:	2701                	sext.w	a4,a4
  ab2d14:	0ff0000f          	fence
  ab2d18:	1005a6af          	lr.w	a3,(a1)
  ab2d1c:	00f69563          	bne	a3,a5,ab2d26 <_ZN14ShenandoahHeap26conc_update_with_forwardedI9narrowOopEEvPT_+0x72>
  ab2d20:	18e5a62f          	sc.w	a2,a4,(a1)
  ab2d24:	fa75                	bnez	a2,ab2d18 <_ZN14ShenandoahHeap26conc_update_with_forwardedI9narrowOopEEvPT_+0x64>
  ab2d26:	0ff0000f          	fence
  ab2d2a:	6422                	ld	s0,8(sp)
```

zifeihan commented:
Hi, When I use this patch, it doesn't compile, and reports the following error msg:
```
/home/zifeihan/jdk/src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:190:27: note: the comparison reduces to ‘(1 >= 8)’
  190 |   STATIC_ASSERT(byte_size >= 8);
/home/zifeihan/jdk/src/hotspot/share/utilities/debug.hpp:289:44: note: in definition of macro ‘STATIC_ASSERT’
   ... (rest of output omitted)
* For target hotspot_variant-server_libjvm_objs_arraycopynode.o:
In file included from /home/zifeihan/jdk/src/hotspot/share/memory/allocation.hpp:29,
                 from /home/zifeihan/jdk/src/hotspot/share/classfile/classLoaderData.hpp:28,
                 from /home/zifeihan/jdk/src/hotspot/share/precompiled/precompiled.hpp:34:
/home/zifeihan/jdk/src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp: In instantiation of ‘T Atomic::PlatformCmpxchg<byte_size>::operator()(volatile T*, T, T, atomic_memory_order) const [with T = unsigned char; long unsigned int byte_size = 1]’:
/home/zifeihan/jdk/src/hotspot/share/runtime/atomic.hpp:1026:40:   required from ‘T Atomic::CmpxchgImpl<T, T, T, typename std::enable_if<std::is_integral<_Tp>::value, void>::type>::operator()(volatile T*, T, T, atomic_memory_order) const [with T = unsigned char; typename std::enable_if<std::is_integral<_Tp>::value, void>::type = void]’
/home/zifeihan/jdk/src/hotspot/share/runtime/atomic.hpp:1002:32:   required from ‘static D Atomic::cmpxchg(volatile D*, U, T, atomic_memory_order) [with D = unsigned char; U = unsigned char; T = unsigned char]’
/home/zifeihan/jdk/src/hotspot/share/oops/methodData.hpp:216:46:   required from here
/home/zifeihan/jdk/src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:190:27: error: static assertion failed: byte_size >= 8
  190 |   STATIC_ASSERT(byte_size >= 8);
/home/zifeihan/jdk/src/hotspot/share/utilities/debug.hpp:289:44: note: in definition of macro ‘STATIC_ASSERT’
  289 | #define STATIC_ASSERT(Cond) static_assert((Cond), #Cond)
      |                                            ^~~~
/home/zifeihan/jdk/src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:190:27: note: the comparison reduces to ‘(1 >= 8)’
  190 |   STATIC_ASSERT(byte_size >= 8);
/home/zifeihan/jdk/src/hotspot/share/utilities/debug.hpp:289:44: note: in definition of macro ‘STATIC_ASSERT’
   ... (rest of output omitted)
* For target hotspot_variant-server_libjvm_objs_assembler.o:
In file included from /home/zifeihan/jdk/src/hotspot/share/memory/allocation.hpp:29,
                 from /home/zifeihan/jdk/src/hotspot/share/classfile/classLoaderData.hpp:28,
                 from /home/zifeihan/jdk/src/hotspot/share/precompiled/precompiled.hpp:34:
/home/zifeihan/jdk/src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp: In instantiation of ‘T Atomic::PlatformCmpxchg<byte_size>::operator()(volatile T*, T, T, atomic_memory_order) const [with T = unsigned char; long unsigned int byte_size = 1]’:
/home/zifeihan/jdk/src/hotspot/share/runtime/atomic.hpp:1026:40:   required from ‘T Atomic::CmpxchgImpl<T, T, T, typename std::enable_if<std::is_integral<_Tp>::value, void>::type>::operator()(volatile T*, T, T, atomic_memory_order) const [with T = unsigned char; typename std::enable_if<std::is_integral<_Tp>::value, void>::type = void]’
/home/zifeihan/jdk/src/hotspot/share/runtime/atomic.hpp:1002:32:   required from ‘static D Atomic::cmpxchg(volatile D*, U, T, atomic_memory_order) [with D = unsigned char; U = unsigned char; T = unsigned char]’
/home/zifeihan/jdk/src/hotspot/share/runtime/atomic.hpp:750:38:   required from ‘T Atomic::PrefetchBitopsUsingCmpxchg::bitop(volatile T*, atomic_memory_order, Op) const [with T = unsigned char; Op = Atomic::PrefetchBitopsUsingCmpxchg::fetch_then_or<unsigned char>(volatile unsigned char*, unsigned char, atomic_memory_order) const::<lambda(unsigned char)>]’
/home/zifeihan/jdk/src/hotspot/share/runtime/atomic.hpp:763:17:   required from ‘T Atomic::PrefetchBitopsUsingCmpxchg::fetch_then_or(volatile T*, T, atomic_memory_order) const [with T = unsigned char]’
/home/zifeihan/jdk/src/hotspot/share/runtime/atomic.hpp:194:53:   required from ‘static T Atomic::fetch_then_or(volatile T*, T, atomic_memory_order) [with T = unsigned char]’
/home/zifeihan/jdk/src/hotspot/share/oops/instanceKlassFlags.hpp:126:58:   required from here
/home/zifeihan/jdk/src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:190:27: error: static assertion failed: byte_size >= 8
  190 |   STATIC_ASSERT(byte_size >= 8);
/home/zifeihan/jdk/src/hotspot/share/utilities/debug.hpp:289:44: note: in definition of macro ‘STATIC_ASSERT’
  289 | #define STATIC_ASSERT(Cond) static_assert((Cond), #Cond)
      |                                            ^~~~
   ... (rest of output omitted)

* All command lines available in /home/zifeihan/jdk/build/linux-riscv64-server-release/make-support/failure-logs.
=== End of repeated output ===

No indication of failed target found.
HELP: Try searching the build log for '] Error'.
HELP: Run 'make doctor' to diagnose build problems.

make[1]: *** [/home/zifeihan/jdk/make/Init.gmk:323: main] Error 2
make: *** [/home/zifeihan/jdk/make/Init.gmk:189: default] Error 2
```

gcc version:
````
[zifeihan@fedora-riscv jdk]$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/riscv64-redhat-linux/13/lto-wrapper
Target: riscv64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,m2,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --enable-libstdcxx-backtrace --with-libstdcxx-zoneinfo=/usr/share/zoneinfo --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-13.2.1-20230728/obj-riscv64-redhat-linux/isl-install --enable-gnu-indirect-function --with-arch=rv64gc --with-abi=lp64d --with-multilib-list=lp64d --build=riscv64-redhat-linux --with-build-config=bootstrap-lto --enable-link-serialization=1
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.2.1 20230728 (Red Hat 13.2.1-1) (GCC)
```

MaxXSoft commented:
@zifeihan 

Sorry for that, I fixed this assertion in the latest commit. Now it should compile. Please let me know if not.

zifeihan commented:
Hi, I tried the latest version and everything works well with the build. And the tier1 test is working fine.

zifeihan commented:
> @zifeihan
> 
> Sorry for that, I fixed this assertion in the latest commit. Now it should compile. Please let me know if not.

@MaxXSoft Thanks for fix that. I tried the latest version and everything works well with the build. And the tier1 test is working fine.

MaxXSoft commented:
> @MaxXSoft Thanks for fix that. I tried the latest version and everything works well with the build. And the tier1 test is working fine.

@zifeihan Happy to hear that. Thanks for your testing!

RealFYang commented on src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:
```diff
@@ -114,6 +120,39 @@ inline T Atomic::PlatformCmpxchg<1>::operator()(T volatile* dest __attribute__((
+  STATIC_ASSERT(4 == sizeof(T));
+
+  T old_value;
+  long flag;
```
To be consistent with the preceding `Atomic::PlatformCmpxchg<1>`, I would suggest change `long flag;` into `uint64_t rc_temp;`

MaxXSoft replied:
Updated.

RealFYang reviewed (changes requested):
Hi, I have several minor comments. BTW: I am curious about whether GCC old versions like GCC-11/12 would work. Can you take a look? Thanks.

MaxXSoft commented on src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:
```diff
@@ -39,6 +39,12 @@
+#elif defined(__GNUC__) && __riscv_xlen <= 32
+#define CORRECT_COMPILER_ATOMIC_SUPPORT
+#endif
+
```
Reason for adding this macro:

* `__atomic_compare_exchange` is correct in Clang, we can directly use it with `uint32_t` for potential optimization opportunities.
* If this bug has been fixed in a later version of GCC, we can enable this macro for it, and leave the macro disabled for other GCC versions.

MaxXSoft commented on src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:
```diff
@@ -114,6 +120,39 @@ inline T Atomic::PlatformCmpxchg<1>::operator()(T volatile* dest __attribute__((
+#ifndef CORRECT_COMPILER_ATOMIC_SUPPORT
+template<>
+template<typename T>
+inline T Atomic::PlatformCmpxchg<4>::operator()(T volatile* dest __attribute__((unused)),
```
Updated.

MaxXSoft commented:
@RealFYang Thanks for your comments!

> BTW: I am curious about why GCC old versions like GCC-11/12 would work. Can you take a look? Thanks.

I'm trying to figure this out. Are you building in `release` mode? `fastdebug` enables `CHECK_UNHANDLED_OOPS` macro, which adds explicit null checks before every oop dereferences, so the bug will not appear.

RealFYang commented:
> @RealFYang Thanks for your comments!
> 
> > BTW: I am curious about why GCC old versions like GCC-11/12 would work. Can you take a look? Thanks.
> 
> I'm trying to figure this out. Are you building in `release` mode? `fastdebug` enables `CHECK_UNHANDLED_OOPS` macro, which adds explicit null checks before every oop dereferences, so the bug will not appear.

Yeah, I am talking about linux-riscv64 release build built with GCC-11/12.

robehn commented:
Hey,

> So the return value should be sign-extended, even if it's unsigned (according to my understand). Here's a simple example: https://godbolt.org/z/n6vfoYPPx. GCC generates `slliw` for the unsigned return value.
> 

This method should, at least often, be inlined, so old_value will not be passed as a return value.
And someone may add additional code in the same method.
Now it seems like the compiler always sign extends every word from inline asm, but I don't think that is a guarantee?
I was told always pass registers to inline asm and not types.
So I suggest having the correct sign extended word type, e.g. int32_t, as old_value.

> > Anyhow can't we just use a int32_t when comparing, and cast that back to the 'uint32_t' on return? (and still use _atomic...)
> 
> Seems we can't. Check this example for more details: https://godbolt.org/z/cv5ejTfxh. I wrote both the inline assembly version and the built-in version CAS, and tested them on the same code snippet. The inline assembly version works well due to the ABI, but the built-in version is still buggy.

Yes, I don't find a good way to force it. It possible using volatile but it involves storing and load from stack.

Thanks, Robbin

MaxXSoft commented:
> Yeah, I am talking about linux-riscv64 release build built with GCC-11/12.

@RealFYang 

Hello! I tried to build linux-riscv64 release with GCC 12, the test result shows that it works. I also checked the dump of `libjvm.so`, `sext.w` does appear in it:

```
  a8d268:	c609                	beqz	a2,a8d272 <_ZN14ShenandoahHeap26conc_update_with_forwardedI9narrowOopEEvPT_+0x60>
  a8d26a:	00a7d7b3          	srl	a5,a5,a0
  a8d26e:	0007869b          	sext.w	a3,a5
  a8d272:	0ff0000f          	fence
  a8d276:	00a75733          	srl	a4,a4,a0
  a8d27a:	1005a7af          	lr.w	a5,(a1)
  a8d27e:	00d79563          	bne	a5,a3,a8d288 <_ZN14ShenandoahHeap26conc_update_with_forwardedI9narrowOopEEvPT_+0x76>
  a8d282:	18e5a62f          	sc.w	a2,a4,(a1)
  a8d286:	fa75                	bnez	a2,a8d27a <_ZN14ShenandoahHeap26conc_update_with_forwardedI9narrowOopEEvPT_+0x68>
  a8d288:	0ff0000f          	fence
  a8d28c:	6422                	ld	s0,8(sp)
```

But it seems that other optimizations interfere with the generated code. If someone changes the source code, the sign extension will disappear due to the GCC bug. Check out this short example simplified from the code of `ShenandoahHeap::conc_update_with_forwarded<narrowOop>`: https://godbolt.org/z/TY84KKx5f.

So we still need this patch on GCC 11/12.

MaxXSoft commented:
Hi @robehn , Thanks for your suggestions!

> Now it seems like the compiler always sign extends every word from inline asm, but I don't think that is a guarantee?
> I was told always pass registers to inline asm and not types.
> So I suggest having the correct sign extended word type, e.g. int32_t, as old_value.

You are right. I tried the inline assembly version CAS on Clang (although we don't enable this patch for Clang), and the parameters don't seem to be sign-extended unless we force it to be: https://godbolt.org/z/fM8n9GMcz.

I added an explicit type cast for `compare_value`, and changed the type of `old_value` to `int32_t` in the latest commit.

robehn reviewed (approved):
Thank you!

I played a bit inline asm and it's actually handled pretty horrible with regards to zero/sign extensions. It seems like in some cases it is not possible to get the compiler to do the correct thing.

RealFYang commented on src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:
```diff
@@ -114,6 +120,44 @@ inline T Atomic::PlatformCmpxchg<1>::operator()(T volatile* dest __attribute__((
+  if (order != memory_order_relaxed) {
+    FULL_MEM_BARRIER;
+  }
+  return old_value;
```
Should we do an explicit type conversion here? Like: `return (T)old_value;`

MaxXSoft replied:
Updated.

RealFYang reviewed (approved):
LGTM. Thanks. The condition for defining macro `CORRECT_COMPILER_ATOMIC_SUPPORT` needs to be improved when the gcc bugfix is backported to gcc branches 11, 12 and 13 [1].

[1] https://gcc.gnu.org/pipermail/gcc-patches/2024-February/646822.html

RealFYang commented:
/sponsor

MaxXSoft commented:
/backport jdk22u

zifeihan commented:
/backport jdk22u
